blockdevice/util/disk/disks.go

```diff
@@ -0,0 +1,298 @@
+		}
+
+		blockSize, _ := strconv.ParseUint(strings.TrimSpace(blockSizeString), 10, 64) //nolint:errcheck
+
+		size *= blockSize
+	}
+
+	diskType := TypeUnknown
+	rotational := readFile(sysblock, dev, "queue/rotational")
+
+	switch {
+	case rotational == "1":
+		diskType = TypeHDD
+	case rotational == "0":
+		diskType = TypeSSD
+	case strings.Contains(dev, "nvme"):
+		diskType = TypeNVMe
+	case strings.Contains(dev, "mmc"):
+		diskType = TypeSD
+	}
+
+	uuid := readFile(sysblock, dev, "uuid")
+	if uuid == "" {
+		uuid = readFile(sysblock, dev, "device/uuid")
+	}
+
+	wwid := readFile(sysblock, dev, "wwid")
+	if wwid == "" {
+		wwid = readFile(sysblock, dev, "device/wwid")
+	}
+
+	serial := readFile(sysblock, dev, "serial")
+	if serial == "" {
+		serial = readFile(sysblock, dev, "device/serial")
+	}
+
+	return &Disk{
+		DeviceName: fmt.Sprintf("/dev/%s", dev),
+		Size:       size,
+		Model:      readFile(sysblock, dev, "device/model"),
+		Name:       readFile(sysblock, dev, "device/name"),
+		Serial:     serial,
+		Modalias:   readFile(sysblock, dev, "device/modalias"),
+		WWID:       wwid,
+		UUID:       uuid,
+		Type:       diskType,
+	}
+}
+
+// Find disk matching provided spec.
+// string parameters may include wildcards.
+func Find(spec *Disk, matchers ...Matcher) (*Disk, error) {
+	disks, err := List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, disk := range disks {
+		if matches, err := Match(disk, spec, matchers...); err != nil {
+			return nil, err
+		} else if matches {
+			return disk, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// Matcher is a function that can handle some custom disk matching logic.
+type Matcher func(disk *Disk) bool
+
+// Match checks if the disk matches the spec.
+// Spec can contain part of the field and strings can contain wildcards.
+// "and" condition is used when this spec is processed.
+func Match(disk, spec *Disk, matchers ...Matcher) (bool, error) {
+	specValue := reflect.ValueOf(spec)
+	if specValue.Kind() == reflect.Ptr {
+		specValue = specValue.Elem()
+	}
+
+	actualValue := reflect.ValueOf(disk)
+	if actualValue.Kind() == reflect.Ptr {
+		actualValue = actualValue.Elem()
+	}
+
+	for _, match := range matchers {
+		if !match(disk) {
+			return false, nil
+		}
+	}
+
+	for i := 0; i < specValue.NumField(); i++ {
+		if specValue.Field(i).IsZero() {
+			continue
+		}
+
+		expected := specValue.Field(i).Interface()
+		actual := actualValue.Field(i).Interface()
+
+		switch e := expected.(type) {
+		case string:
+			if actualString, ok := actual.(string); ok {
+				matched, err := matchPattern(actualString, e)
+				if err != nil {
+					return false, err
+				}
+
+				if !matched {
+					return false, nil
+				}
+			} else {
+				return false, nil
+			}
+
+		default:
+			if actual != expected {
+				return false, nil
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func matchPattern(in, value string) (bool, error) {
+	re, err := regexp.Compile(
+		fmt.Sprintf("^%s$", strings.ReplaceAll(value, "*", ".*")),
```

Comment:
this seems to be wrong in many ways, at the very least we should quote meta characters https://pkg.go.dev/regexp#QuoteMeta, but still preserving `*` will be tricky

Reply:
oh right, should escape regexp 🤦‍♂️ 

Ok, I just pulled a specialized library for glob matching instead. It's tiny anyways.